### ISSUE TITLE
Feature Request: Ranged Download for BAM and VCF Files

### ISSUE THREAD

**Issue by berntpopp** [LaborBerlin/varvis-download#22]
### Description:
Implement support for ranged downloads of BAM and VCF files using external tools like `samtools` and `tabix`. The new feature should allow users to specify genomic regions for download, making the process more efficient when only certain regions are required. This feature will involve creating a new utility script to handle the ranged requests and file management.

### Acceptance Criteria:
1. **Tool Verification**:
   - The script should first check if the required external tools (`samtools` and `tabix`) are available on the system.
   - Check if the tools meet the minimal required version. If the tools are not available or the versions are insufficient, the script should log an error and exit gracefully.

2. **Performing Ranged Requests**:
   - After downloading the respective index files locally, the script should perform the ranged requests using:
     - `samtools` for BAM files.
     - `tabix` for VCF files.
   - The script should support commands like the following example for `samtools`:
     \`\`\`bash
     samtools view '<URL>' chr1:1-100000
     \`\`\`
   - The output should be the specified genomic region.

3. **File Naming**:
   - The resulting files should be renamed by adding the genomic range to the file name in a normalized form (e.g., `chr1_1_100000` before the file extension).
   - For example:
     - Input file: `LB24-ONTCCMJH308-ready.bam`
     - Output file after ranged request: `LB24-ONTCCMJH308-ready.chr1_1_100000.bam`

4. **Index File Handling**:
   - Before performing the ranged requests, ensure the necessary index files (`.bai` or `.tbi`) are downloaded and available locally.

5. **Utility Script**:
   - Create a new utility script to handle the ranged requests.
   - The script should handle downloading the index files and using the appropriate external commands for processing.
   - The script should log the process, including tool checks, the region being downloaded, and the final file output.

### Sample Command Example:
   \`\`\`bash
   samtools view '<URL>' chr1:1-100000
   \`\`\`

### Additional Notes:
- Ensure the process is well-logged and user-friendly.
- The feature should be integrated in a way that it can be easily invoked within the existing system.
- Implement error handling for situations where tools are missing or the ranged request fails.

### Resources:
- **samtools documentation**: [https://www.htslib.org/doc/samtools.html](https://www.htslib.org/doc/samtools.html)
- **tabix documentation**: [http://www.htslib.org/doc/tabix.html](http://www.htslib.org/doc/tabix.html)

---

### Tasks:
1. Check for required tools (`samtools`, `tabix`) and versions.
2. Implement ranged requests for BAM and VCF files.
3. Handle renaming of files with the specified range.
4. Ensure index files are downloaded before performing requests.
5. Create detailed logging for all processes.

**Comment by your-highness** [comment]
I am reopening this.

We need to deal with the fact that ranged downloads currently result in 2 downloads of bam files which counts towards the varvis aws contigency (cost risk). I think this is related to the current samtools version used. However, I can not provide much more detail

Can you please comment if this is still an issue with the current implementation of varvis-download, @berntpopp ?

CC @AxelMacFoly @selyasar

**Comment by berntpopp** [comment]
I can not reproduce this. Please provide details on the setup (tool version, samtools version) and ideally logs.

**Comment by your-highness** [comment]
> I can not reproduce this. Please provide details on the setup (tool version, samtools version) and ideally logs.

@AxelMacFoly can provide details

**Comment by AxelMacFoly** [comment]
I had a phone call with @berntpopp and I gave him some details.

For documentation, we are using the following samtools version:
>samtools 1.20
>Using htslib 1.23

**Comment by berntpopp** [comment]
Problem seems to be scattered throughout versions and combinations of samtools and htslb.
@AxelMacFoly : Could you pelase try with the newest version of both tools and tabix and tell me if you still see the same phenomenon?
If so this is a larger fix of replacing samtools/tabix/htslib with a JS native implementation or dropping ranged request support or warning of "virtual" traffic.

**Comment by berntpopp** [comment]
The upstream situation is now clearer:

- HTSlib [#1670](https://github.com/samtools/htslib/issues/1670) was fixed by [PR #1958](https://github.com/samtools/htslib/pull/1958) in **HTSlib 1.23**, but only for its native `s3://` backend, which now makes bounded chunk requests.
- Varvis currently supplies pre-signed `https://...` S3 URLs. Samtools/tabix route these through HTSlib's generic libcurl backend, which still makes open-ended requests (`Range: bytes=N-`) through HTSlib 1.24.
- The corresponding fix for pre-signed HTTPS URLs is [HTSlib PR #1998](https://github.com/samtools/htslib/pull/1998), but it remains open and unmerged. Upgrading alone therefore does not yet guarantee safe egress accounting for the current Varvis path.

A useful controlled test is whether the API-provided URL works through HTSlib's native S3 handler: `s3://`, `s3+http://`, and especially by replacing the leading `https://` with `s3+https://` while preserving the complete host/path/query. This may fail if HTSlib reinterprets or re-signs the pre-signed request, so it should not be assumed to work.

BAM and VCF/tabix should be tested separately with HTSlib >=1.23.2. S3/proxy logs should show whether requests change from `bytes=N-` to bounded `bytes=N-M` and whether accounted bytes fall. Pre-signed URLs must be redacted, and `s3+http://` should only be tried in an isolated non-sensitive test.

**Comment by berntpopp** [comment]
A possible tool-side workaround while waiting for the HTSlib HTTPS fix is a **local bounded-range adapter**:

```text
samtools/tabix -> short-lived localhost adapter -> unchanged presigned HTTPS URL
```

When HTSlib asks the adapter for an open-ended range such as `bytes=X-`, the adapter would fetch the S3 object upstream in bounded chunks (for example 1 MiB), then stream those chunks as one continuous response to HTSlib. When HTSlib closes the response, the adapter aborts the current upstream request, limiting excess S3 transfer to roughly one chunk. Already-bounded requests can be forwarded directly.

This retains samtools/tabix for BAM/BAI/VCF parsing, keeps the original presigned URL and signature untouched, and could use the existing `undici` dependency. It is much smaller and safer than attempting a JavaScript port of HTSlib.

A prototype would still need to verify byte-exact `HEAD`/`Range`/`Content-Range` handling, cancellation and backpressure, URL expiry/refresh, proxy support, loopback-only security, and behavior for both samtools and tabix. If that validation succeeds, this could serve as an experimental workaround until the upstream libcurl backend supports bounded indexed reads.

**Comment by your-highness** [comment]
Any update here?

**Comment by AxelMacFoly** [comment]
Hi @berntpopp,

could you estimate when this issue could realistically be implemented, possibly first on a development branch, and how much time you would expect the implementation to take?

Once there is a testable version, I can run the before/after downloads on our side, request the daily total download statistics from Varvis, and document the results here in the issue, including whether the adapter actually reduces the reported download volume.

Thanks!

**Comment by berntpopp** [comment]
Depends on upstream libraries completely, didn't check again if they implemented the fix

**Comment by berntpopp** [comment]
TL;DR: As of today's check, the issue remains unfixed upstream (HTSlib 1.24 still issues open-ended range requests for presigned HTTPS URLs and PR #1998 is stalled), so we will need to implement our own bounded-range adapter.

**Comment by berntpopp** [comment]
Hi @AxelMacFoly,

The bounded-range reverse proxy is now implemented, fully tested, and ready for testing on a development branch in **PR #157**: https://github.com/LaborBerlin/varvis-download/pull/157

### Summary of the implementation
- **100% Self-Contained**: Runs an ephemeral in-process loopback proxy on `127.0.0.1` using Node.js built-ins. No external proxy daemon or new npm dependencies required.
- **Enabled by Default**: Activated automatically (`--bounded-range-proxy`, default `true`). Can be disabled via `--no-bounded-range-proxy` or tuned via `--bounded-range-chunk-size` (default `2097152` [2 MiB]).
- **Future-Proof Guard**: Inspects local `samtools` and `tabix` versions. If an updated HTSlib ($\ge 1.25.0$) with native bounded HTTPS range support is detected, the proxy is automatically bypassed unless explicitly enforced.
- **Benchmark Savings**: In synthetic real-data tests against 1000 Genomes S3 BAMs (~17.28 GB), in-flight egress dropped from up to ~34.5 GB down to $\le 4\text{ MiB}$ (**>99.98% reduction**).

### How to test the branch
```bash
git fetch origin feature/bounded-range-proxy
git checkout feature/bounded-range-proxy
npm install

# 1. Ranged download with proxy active (default):
./varvis-download.cjs download -u <user> -p <pass> -t <target> -a <analysisId> -r <chr:start-end>

# 2. Before/after comparison download with proxy bypassed:
./varvis-download.cjs download -u <user> -p <pass> -t <target> -a <analysisId> -r <chr:start-end> --no-bounded-range-proxy
```

Could you please benchmark and test this branch with real clinical data on your side, and check with Varvis to confirm whether the daily download statistics reflect the reduction in traffic?

CC @your-highness

**Comment by AxelMacFoly** [comment]
@berntpopp @your-highness 
I successfully ran the tests on the last weekend and send today morning a request for the aws traffic to the varvis support. Normally they respond within a day on such a request. As stated last week, I will report the outcome here when available.
